### PR TITLE
Add variable specific heat functionality for multispecies Tpg rebase

### PR DESCRIPTION
**Pull request by madpeck3**
New species section in the input deck used when the eos (in the materials section) is set to "thermallyperfectgas." The parameters necessary for specifying different thermodynamic properties for each species are set in the new species section.

Regression test added is the multispecies sod problem using the "thermallyperfectgas" eos.

- - -
This change is [Reviewable](https://reviewable.io/reviews/quinoacomputing/quinoa/631)

### CODE REVIEW

**Review by adityakpandare — CHANGES_REQUESTED** [review]
Great work, @madpeck3! I have some comments, but otherwise looks good!

> Reviewed 16 of 63 files at r1.
> *[Reviewable](https://reviewable.io/reviews/quinoacomputing/quinoa/631#-:-ODCM_Vm2mG7wa6xa606:b-16sqni)* status: 16 of 63 files reviewed, 6 unresolved discussions (waiting on @madpeck3)

___
*[`src/PDE/EoS/ThermallyPerfectGas.cpp` line 149 at r1](https://reviewable.io/reviews/quinoacomputing/quinoa/631#-ODCM1vK2kyCuWKRsfTG:-ODCM1vK2kyCuWKRsfTH:btxicxp) ([raw file](https://github.com/quinoacomputing/quinoa/blob/f886a9ad8bef872b00e270a22b46ecb63fd58184/src/PDE/EoS/ThermallyPerfectGas.cpp#L149)):*
> ```cpp
> 
>   tk::real temp = pr / (rho * R);
>   tk::real e = R * (-m_cp_coeff[0] * pow(temp, -1) +
> ```

Let's use the `std::pow` function instead. The std library pow function is type-safe. Let's do that for all `pow` functions in this class.
___
*[`src/PDE/EoS/ThermallyPerfectGas.cpp` line 150 at r1](https://reviewable.io/reviews/quinoacomputing/quinoa/631#-ODCMMmJ6Q4vYF5tk8Dt:-ODCMMmJ6Q4vYF5tk8Du:b-j5l4tk) ([raw file](https://github.com/quinoacomputing/quinoa/blob/f886a9ad8bef872b00e270a22b46ecb63fd58184/src/PDE/EoS/ThermallyPerfectGas.cpp#L150)):*
> ```cpp
>   tk::real temp = pr / (rho * R);
>   tk::real e = R * (-m_cp_coeff[0] * pow(temp, -1) +
>       m_cp_coeff[1] * log(temp) + (m_cp_coeff[2] - 1) * temp +
> ```

Same for `log`- let's use `std::log` instead.
___
*[`src/PDE/EoS/ThermallyPerfectGas.cpp` line 187 at r1](https://reviewable.io/reviews/quinoacomputing/quinoa/631#-ODCLTMhDve_v58Bu1d-:-ODCLTMhDve_v58Bu1d0:b6w9d8a) ([raw file](https://github.com/quinoacomputing/quinoa/blob/f886a9ad8bef872b00e270a22b46ecb63fd58184/src/PDE/EoS/ThermallyPerfectGas.cpp#L187)):*
> ```cpp
>   tk::real tol = 1e-8; // Stopping condition
>   tk::real err = 1e8;
>   while (err > tol) {
> ```

This can be dangerous. Let's replace this with a for-loop instead with a fixed number of iterations (see `JWL::bisection`).
___
*[`src/PDE/EoS/ThermallyPerfectGas.cpp` line 204 at r1](https://reviewable.io/reviews/quinoacomputing/quinoa/631#-ODCLnMHAJSlifQyMGZ2:-ODCLnMHAJSlifQyMGZ3:buqwdba) ([raw file](https://github.com/quinoacomputing/quinoa/blob/f886a9ad8bef872b00e270a22b46ecb63fd58184/src/PDE/EoS/ThermallyPerfectGas.cpp#L204)):*
> ```cpp
>       if (k == 2) fp_T += -1;
>       power += 1;
>     }
> ```

What root-finder is this? Could you mention that in a comment?
___
*[`tests/regression/inciter/multispecies/SodShocktube/sod_shocktube_p0p1.lua` line 76 at r1](https://reviewable.io/reviews/quinoacomputing/quinoa/631#-ODCKH651IQl5FO5pqjO:-ODCKH651IQl5FO5pqjP:b-c4wr4b) ([raw file](https://github.com/quinoacomputing/quinoa/blob/f886a9ad8bef872b00e270a22b46ecb63fd58184/tests/regression/inciter/multispecies/SodShocktube/sod_shocktube_p0p1.lua#L76)):*
> ```Lua
>       "z-velocity"
>     },
>     elemalias = {
> ```

I suggest not using the 'elemalias' capability, since it will be deprecated soon. When that's done, the regression test baselines will have to be updated with the variables names in the 'elemvar'. Instead, let's remove the elemalias, and use only the elemvar right now.
___
*[`tests/regression/inciter/multispecies/SodShocktube/exodiff_dg.cfg` line 11 at r1](https://reviewable.io/reviews/quinoacomputing/quinoa/631#-ODCKpnR6Xu1mP6YTJB9:-ODCKpnR6Xu1mP6YTJBA:bk3rqzl) ([raw file](https://github.com/quinoacomputing/quinoa/blob/f886a9ad8bef872b00e270a22b46ecb63fd58184/tests/regression/inciter/multispecies/SodShocktube/exodiff_dg.cfg#L11)):*
> ```INI
> 	x-velocity_numerical
> 	y-velocity_numerical
> 	z-velocity_numerical
> ```

All these variable names will have to be updated once you remove the elemalias vector in the control files, for the regression tests to pass.

File: src/PDE/EoS/ThermallyPerfectGas.cpp [hunk 1]
```diff
@@ -180,31 +180,41 @@ ThermallyPerfectGas::temperature(
...
-  while (err > tol) {
-    tk::real f_T = R * (-m_cp_coeff[0] * pow(temp, -1) +
-      m_cp_coeff[1] * log(temp) + (m_cp_coeff[2] - 1) * temp +
-      m_cp_coeff[3] * pow(temp, 2) / 2 +
-      m_cp_coeff[4] * pow(temp, 3) / 3 +
-      m_cp_coeff[5] * pow(temp, 4) / 4 +
-      m_cp_coeff[6] * pow(temp, 5) / 5 + m_cp_coeff[7]) - e;
+  std::size_t maxiter = 1000;
+  std::size_t i(0);
+  while (i < maxiter) {
```
Comment (madpeck3) [inline src/PDE/EoS/ThermallyPerfectGas.cpp:189]: Added this based on `JWL::Bisection`

File: src/PDE/EoS/ThermallyPerfectGas.cpp [hunk 2]
```diff
@@ -180,31 +180,41 @@ ThermallyPerfectGas::temperature(
...
     fp_T = fp_T * R;
 
     temp = temp - f_T / fp_T;
+
+    if (err <= tol) break;
+    i++;
+    if ( i == maxiter ) {
+      Throw("ThermallyPerfectGas Newton's Method for temperature failed to converge after iterations "
+      + std::to_string(i));
+    }
```
Comment (madpeck3) [inline src/PDE/EoS/ThermallyPerfectGas.cpp:217]: Also added based on `JWL::Bisection`

**Comment by madpeck3** [comment]
The changes pushed should address all the comments by @adityakpandare

**Review by adityakpandare — APPROVED** [review]
Thanks, @madpeck3. :lgtm:

> Reviewed 3 of 47 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/quinoacomputing/quinoa/631#-:-ODqW04ZAc9YZVf-mqi-:b-vyembi)* status: 16 of 63 files reviewed, 8 unresolved discussions (waiting on @madpeck3)

**Review by madpeck3 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/quinoacomputing/quinoa/631)* status: 16 of 63 files reviewed, 6 unresolved discussions (waiting on @adityakpandare)

___
*[`src/PDE/EoS/ThermallyPerfectGas.cpp` line 149 at r1](https://reviewable.io/reviews/quinoacomputing/quinoa/631#-ODCM1vK2kyCuWKRsfTG:-ODrOLzQ5muhOAk_04CA:b-896fix) ([raw file](https://github.com/quinoacomputing/quinoa/blob/f886a9ad8bef872b00e270a22b46ecb63fd58184/src/PDE/EoS/ThermallyPerfectGas.cpp#L149)):*
<details><summary><i>Previously, adityakpandare (Aditya Pandare) wrote…</i></summary><blockquote>

Let's use the `std::pow` function instead. The std library pow function is type-safe. Let's do that for all `pow` functions in this class.
</blockquote></details>

Done.
___
*[`src/PDE/EoS/ThermallyPerfectGas.cpp` line 150 at r1](https://reviewable.io/reviews/quinoacomputing/quinoa/631#-ODCMMmJ6Q4vYF5tk8Dt:-ODrOMFa0Pxi95iG63eq:b-896fix) ([raw file](https://github.com/quinoacomputing/quinoa/blob/f886a9ad8bef872b00e270a22b46ecb63fd58184/src/PDE/EoS/ThermallyPerfectGas.cpp#L150)):*
<details><summary><i>Previously, adityakpandare (Aditya Pandare) wrote…</i></summary><blockquote>

Same for `log`- let's use `std::log` instead.
</blockquote></details>

Done.
___
*[`src/PDE/EoS/ThermallyPerfectGas.cpp` line 187 at r1](https://reviewable.io/reviews/quinoacomputing/quinoa/631#-ODCLTMhDve_v58Bu1d-:-ODrOMhG1ZJk0nAJ-QFU:b-896fix) ([raw file](https://github.com/quinoacomputing/quinoa/blob/f886a9ad8bef872b00e270a22b46ecb63fd58184/src/PDE/EoS/ThermallyPerfectGas.cpp#L187)):*
<details><summary><i>Previously, adityakpandare (Aditya Pandare) wrote…</i></summary><blockquote>

This can be dangerous. Let's replace this with a for-loop instead with a fixed number of iterations (see `JWL::bisection`).
</blockquote></details>

Done.
___
*[`src/PDE/EoS/ThermallyPerfectGas.cpp` line 204 at r1](https://reviewable.io/reviews/quinoacomputing/quinoa/631#-ODCLnMHAJSlifQyMGZ2:-ODrON4C21GcAftFwJPf:b-896fix) ([raw file](https://github.com/quinoacomputing/quinoa/blob/f886a9ad8bef872b00e270a22b46ecb63fd58184/src/PDE/EoS/ThermallyPerfectGas.cpp#L204)):*
<details><summary><i>Previously, adityakpandare (Aditya Pandare) wrote…</i></summary><blockquote>

What root-finder is this? Could you mention that in a comment?
</blockquote></details>

Done.
___
*[`tests/regression/inciter/multispecies/SodShocktube/exodiff_dg.cfg` line 11 at r1](https://reviewable.io/reviews/quinoacomputing/quinoa/631#-ODCKpnR6Xu1mP6YTJB9:-ODrOOyQ9x3RD4IgWyC2:b-896fix) ([raw file](https://github.com/quinoacomputing/quinoa/blob/f886a9ad8bef872b00e270a22b46ecb63fd58184/tests/regression/inciter/multispecies/SodShocktube/exodiff_dg.cfg#L11)):*
<details><summary><i>Previously, adityakpandare (Aditya Pandare) wrote…</i></summary><blockquote>

All these variable names will have to be updated once you remove the elemalias vector in the control files, for the regression tests to pass.
</blockquote></details>

Done.
___
*[`tests/regression/inciter/multispecies/SodShocktube/sod_shocktube_p0p1.lua` line 76 at r1](https://reviewable.io/reviews/quinoacomputing/quinoa/631#-ODCKH651IQl5FO5pqjO:-ODrOQ0l2sMySYg_CoEa:b-896fix) ([raw file](https://github.com/quinoacomputing/quinoa/blob/f886a9ad8bef872b00e270a22b46ecb63fd58184/tests/regression/inciter/multispecies/SodShocktube/sod_shocktube_p0p1.lua#L76)):*
<details><summary><i>Previously, adityakpandare (Aditya Pandare) wrote…</i></summary><blockquote>

I suggest not using the 'elemalias' capability, since it will be deprecated soon. When that's done, the regression test baselines will have to be updated with the variables names in the 'elemvar'. Instead, let's remove the elemalias, and use only the elemvar right now.
</blockquote></details>

Done.

**Review by adityakpandare — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/quinoacomputing/quinoa/631#-:-ODrQ7KN6aMxI6dJHWNL:bnfp4nl)* status: 16 of 63 files reviewed, all discussions resolved

**Review by adityakpandare — APPROVED** [review]
> Reviewed 2 of 63 files at r1, 44 of 47 files at r2, 1 of 1 files at r3.
> *[Reviewable](https://reviewable.io/reviews/quinoacomputing/quinoa/631)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @madpeck3)

**Pull request merged**